회원에 대한 후기를 작성하는 API

sujeong11 opened the pull request:
close #38 

## 작업 사항
- 모임이 끝난 후 모임에 참여한 회원에 대해 후기를 작성하는 API 구현
- 회원이나 모임에 대해 id 값만 필요한 것 같아 Dto에 Long id만 받도록 구현했지만 우기님 피드백을 받고 아래와 같이 변경 (잘 한 건지 모르겠다..)
  - receiverId를 받아 회원을 조회해 MemberDto 생성
  - meetId를 받아 모임과 주소를 조회해 MeetDto 생성
  - 만든 Dto들을 인자로 넣어주어 ReviewDto로 생성하고, toEntity()를 사용해 Point와 Review 테이블에 데이터를 저장한다.

## 참고 자료

- 참고자료 (링크 등)

## 체크리스트

- [x]  PR의 제목은 팀 내 규칙을 준수하여 알맞게 작성하였는가?
- [x]  Merge 하는 브랜치가 올바른가? (`main` branch에 실수로 PR 생성 금지)
- [x]  팀의 코딩 컨벤션을 준수하는가?
- [x]  PR과 관련없는 변경사항이 들어가지는 않았는가?
- [x]  내 코드에 대한 자기 검토가 되었는가?
- [x]  Reviewers, Assignees, Lables, Project, Milestone은 적절하게 선택하였는가?
- [x]  관련한 issue를 닫아야 하는지 점검해보고 적용했는가?

sujeong11 committed:
feat: #38 리뷰 글 작성 API
- senderId는 인증 기능이 머지가 된 후 로그인한 사용자로 수정할 예정
- receiverId를 받아 회원을 조회해 MemberDto 생성
- meetId를 받아 모임과 주소를 조회해 MeetDto 생성
- ReviewCreateRequest로attendance, passion, contactSpeed 값을 가지는 point와 리뷰 내용인 content를 가지며 point는 PointRequest 타입으로 받아 toDto()로 PointDto로 생성
이렇게 만든 Dto들을 인자로 넣어주어 ReviewDto로 생성하고, toEntity()를 사용해 Point와 Review 테이블에 데이터를 저장한다.

sujeong11 committed:
Merge branch 'develop' into feature/#38-create-review-post

sujeong11 committed:
feat: #38 로그인된 사용자 아이디 사용 및 후기 API 로직 변경

sujeong11 committed:
feat: #38 swagger 문서화 추가

Wo-ogie commented on src/main/java/com/umc/withme/dto/point/PointRequest.java:
```diff
@@ -0,0 +1,25 @@
+@Getter
+public class PointRequest {
+
+    @NotNull
+    private Integer attendance;
+
+    @NotNull
+    private Integer passion;
+
+    @NotNull
+    private Integer contactSpeed;
```
Null checking만 하기보다는 최소, 최댓값에 대한 검증도 있으면 좋을 것 같습니다. `@Min`, `@Max` annotation으로 가능하고, 관련 정보는 구글링 하시거나 [Spring Validation Annotation 정리](https://hyeran-story.tistory.com/81)를 참고해보셔도 괜찮을 것 같아요.

Wo-ogie commented on src/main/java/com/umc/withme/dto/review/ReviewCreateRequest.java:
```diff
@@ -0,0 +1,20 @@
+    @NotNull
+    private PointRequest point;
+
+    @NotBlank
+    private String content;
```
저희 리뷰 최대 글자 수가 제한되어 있으므로 여기서 한 번 검증해도 괜찮을 것 같습니다?
`@Length` annotation으로 validation이 가능합니다.

Wo-ogie commented on src/main/java/com/umc/withme/exception/ExceptionType.java:
```diff
@@ -42,6 +44,10 @@ public enum ExceptionType {
     // Member
     NICKNAME_NOT_FOUND_EXCEPTION(1400, "해당 닉네임을 가지는 회원이 없습니다.", NicknameNotFoundException.class),
     EMAIL_NOT_FOUND_EXCEPTION(1401, "해당 이메일을 갖는 회원을 찾을 수 없습니다.", EmailNotFoundException.class),
+    MEMBER_ID_NOT_FOUND_EXCEPTION(1402, "해당 아이디를 가지는 회원이 없습니다.", MemberIdNotFoundException.class),
+
+    // Meet
+    MEET_ID_NOT_FOUND_EXCEPTION(2400, "해당 아이디를 가지는 모임이 없습니다.", MeetIdNotFoundException.class),
```
이거.... 아이디(id)가 프론트엔드 입장에서 받았을 때 PK가 아닌 로그인 시 필요한 id로 오해할 수 있을 것 같아서... 혹시 더 좋은 표현은 없을까요?
저는 그래서 중간에 할 때 "일치하는 회원을 찾을 수 없습니다." 정도로만 했었는데, 고민이 되는 부분입니당....

sujeong11 replied:
`MEMBER_PK_ID_NOT_FOUND_EXCEPTION`나 `MEMBER_PK_NOT_FOUND_EXCEPTION`로 하는 건 어떨까요?

Wo-ogie replied:
그 방법도 좋을 것 같은데 저는 `errorMessage`를 말씀드린 거였어요! 저 부분을 저희가 봤을때는 무슨 문제인지 바로 알 수 있는데 아무래도 아이디라는 표현이 프론트엔드 개발자나 사용자가 딱 읽었을 때 고유 식별자 뿐만 아니라 계정 아이디라는 의미와 혼동이 있을 수 있지 않을까? 싶었던 거에요.
저도 어떻게 해야 좋을지는 딱히 생각이 안나서 그냥 의견을 한 번 여쭤봤습니다 😊

sujeong11 replied:
아! "해당 아이디(PK)를 가지는 회원이 없습니다."라고 하면 이해하실 수 있지 않을까요?

Wo-ogie replied:
좋습니당

Wo-ogie commented on src/main/java/com/umc/withme/repository/MeetAddressRepository.java:
```diff
@@ -6,4 +6,6 @@
 
 public interface MeetAddressRepository extends JpaRepository<MeetAddress, Long> {
+
+    List<MeetAddress> findByMeet_Id(Long meetId);
```
이거 method name에서 단건조회인지 전체조회인지 명시해주는게 더 좋을 것 같아요.
`findAllByMeet_Id`로 method name을 사용하면 어떨까요?

Wo-ogie replied:
앗 이거 파크가 한 게 아니네요 ㅋㅋㅋ;;
에이치가 알아서 잘 수정해서 올리실테니 이 리뷰는 무시하셔도 됩니다!

Wo-ogie commented on src/main/java/com/umc/withme/service/ReviewService.java:
```diff
@@ -0,0 +1,54 @@
+    public Long create(Long senderId, Long receiverId, Long meetId, ReviewCreateRequest requestDto) {
+
+
+        Member sender = memberRepository.findById(senderId).orElseThrow(UnauthorizedException::new);
```
그 로그인한 사용자 정보 못받아오면 따로 예외처리하는 부분이 있어서요.
여기에서는 `MemberIdNotFoundException`으로 던지셔도 될 것 같습니다.

Wo-ogie reviewed (commented):
`PointRequest`, `ReviewCreateResponse`, `ReviewCreateRequest`에도 API 명세 작성해주시면 좋을 것 같습니다.
Example 값이랑 설명 정도만요! `@Schema`로 작성하실 수 있고, 제가 한 거 참고하시면 될 것 같습니당.

sujeong11 committed:
feat: #38 validation 추가
포인트를 입력받는 3개 필드에 대해 @Max와 @Min
후기 내용을 입력받는 content에 @Length

sujeong11 committed:
feat: #38 swagger 문서화 추가
요청 / 응답 데이터인 PointRequest, ReviewCreateRequest, ReviewCreateResponse에 대한 예시와 설명 API 명세에 추가

sujeong11 committed:
style: #38 에러처리 수정
- id로 특정 회원이나 모임을 조회 실패 시 발생하는 모호한 에러메시지 수정
해당 아이디를 가지는 회원이 없습니다. => 해당 아이디(PK)를 가지는 회원이 없습니다.
- 후기 작성할 때, 로그인한 회원 id로 회원 조회 실패 시 UnauthorizedException가 발생하도록 했지만, 로그인한 사용자를 받아오지 못하는 에러처리 부분이 존재하므로 MemberIdNotFoundException로 변경함.

sujeong11 committed:
style: #38 불필요한 API 명세 제거

Wo-ogie commented on src/main/java/com/umc/withme/controller/ReviewController.java:
```diff
@@ -32,8 +32,7 @@ public class ReviewController {
-            @ApiResponse(responseCode = "200", description = "OK"),
-            @ApiResponse(responseCode = "404", description = "1402: <code>id</code>에 해당하는 회원이 없는 경우 \t\n 2400: <code>id</code>에 해당하는 모임이 없는 경우", content = @Content)
+            @ApiResponse(responseCode = "200", description = "Success")
     })
```
이거 성공 case만 두실거면 그냥 `@ApiResponses` 아예 지우셔도 될 것 같아요. 기본값으로 200은 자동으로 등록해줘서

Wo-ogie commented on src/main/java/com/umc/withme/dto/review/ReviewCreateRequest.java:
```diff
@@ -4,17 +4,19 @@
 
-    @NotNull
+    @Valid
     private PointRequest point;
```
`ReviewCreateRequest`랑 `PointRequest`에 `@NoArgsConstructor(access = AccessLevel.PROTECTED)` 안붙여도 되나요?

`@RequestBody`로 입력받는 객체는 default constructor가 있어야 하는 걸로 알고 있었어서 테스트 해봤는데 저는 에러떠서요. 
https://zkdlu.github.io/2021-10-07/Spring-boot-%EA%B0%9D%EC%B2%B4-%EB%B0%94%EC%9D%B8%EB%94%A9/
https://velog.io/@wisdom08/cannot-deserialize-from-Object-value-%EC%97%90%EB%9F%AC

파크 환경에서는 괜찮은가요?

sujeong11 replied:
방금 테스트 다시 해봤는데 저는 잘 동작해요.. 그래서 잘못 적은지도 몰랐네요😅

아래 글 보니깐 com.fasterxml.jackson.module:jackson-module-parameter-names:2.13.4.2 덕분인 것 같기도 하네욥
근데 왜 저는 잘 동작하고 우기는 동작 안하는지 모르겠네요
https://bbbicb.tistory.com/46
 https://juhi.tistory.com/68
 
 일단, @NoArgsConstructor(access = AccessLevel.PROTECTED)로 붙여두겠습니다!

Wo-ogie replied:
아하 개발환경이 달라서 그런 것 같네요 ㅠ 뭐가 문제인지 대충 알 것 같은데... 일단은 없어도 괜찮은 것 같네요?
그럼 일단 붙여주시고 이건 제가 좀 환경 테스트 해보고 나중에 다시 말씀드리겠습니다. 감사합니다! 😊

Wo-ogie replied:
확인해보니까 intellij에서 build and run을 하는 build tool을 intellij로 설정하니 error가 발생하네요 ㅠㅠ
첨부해주신 글 참고해서 build tool을 gradle로 설정하니 spring-web dependency에 포함된 com.fasterxml.jackson.module:jackson-module-parameter-names에 의해 잘 입력이 되는 것 같아요.
앞으로는 default constructor 신경쓰지 않고 하셔도 될 것 같습니다!!

Wo-ogie replied:
확인해보니까 intellij에서 build and run을 하는 build tool을 intellij로 설정하니 error가 발생하네요 ㅠㅠ
첨부해주신 글 참고해서 build tool을 gradle로 설정하니 spring-web dependency에 포함된 `jackson-module-parameter-names`에 의해 잘 입력이 되는 것 같아요.

그래서 build tool을 gradle과 intellij 중 어느 것으로 선택하는 것이 좋을지 고민을 해봤는데 다음 글이 도움이 되었습니다. 궁금하시면 한 번 읽어보셔도 좋을 것 같아요!
https://pamyferret.tistory.com/m/62

결국 intellij로 build 했을 때 왜 `jackson-module-parameter-names`가 잘 동작하지 않는지는 정확히 모르겠지만, 저도 build and run은 gradle로 동작하도록 설정하고 사용하려고 합니다. **앞으로는 default constructor 신경쓰지 않고 하셔도 될 것 같습니다!!**

sujeong11 replied:
둘의 차이를 확실히 알 수 있네요! 다른 팀원 분들도 읽어보면 좋은 글인 것 같아요👍

Wo-ogie commented on src/main/java/com/umc/withme/controller/ReviewController.java:
```diff
@@ -0,0 +1,52 @@
+
+import javax.validation.Valid;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api")
+public class ReviewController {
```
Controller class type level에 `@Tag` annotation도 작성해서 간단하게 제목 붙여주시면 좋을 것 같습니당.

sujeong11 committed:
style: #38 API 명세 수정

sujeong11 committed:
style: #38 생성자 어노테이션 제거

Wo-ogie commented:
Merge 충돌난 거 수정하시면 approve 하도록 하겠습니다!

sujeong11 committed:
Merge branch 'develop' into feature/#38-create-review-post

Wo-ogie reviewed (changes requested):
이거 혹시 conflict resolve 하시고 실행해보셨나요..? Compile error 뜨는데...

Exception에 default constructor 추가하셔야 할 듯 합니다?

sujeong11 committed:
fix: #38 충돌 해결 후 compile error 해결
- 기본 생성자가 없어 발생하는 에러 발생해 기본 생성자 추가

Wo-ogie commented:
혹시 문제 다 해결되시고 review 기다리시는 거라면 `Re-request review` 한 번 눌러주세요!

sujeong11 committed:
Merge branch 'develop' into feature/#38-create-review-post
